<font color='blue'>Here I will document my method for cleaning my data files, which were taken from Kaggle's dataset “Comments on articles published in the New York Times” (https://www.kaggle.com/aashita/nyt-comments).

This code is broken into two sections - one cleans the data files with articles and one cleans the data files with comments. Here I test my cleaning using only a single file, but by concatenating data files together, I can run this code on the entire dataset.</font> 

In [40]:
import pandas as pd
import numpy as np
import nltk
#nltk.download('wordnet')
import re
sent_token = nltk.sent_tokenize
import csv  
from nltk import sent_tokenize, word_tokenize, pos_tag
import re
from sklearn.feature_extraction.text import CountVectorizer
wpt = nltk.WordPunctTokenizer()
stop_words = nltk.corpus.stopwords.words('english')

#For all data - train = pd.read_csv("/root/Springboard/Data/cleaning/allArticles.csv")
train = pd.read_csv(r"C:\Users\msteele9\Documents\Springboard\Springboard\Data\ArticlesApril2017.csv")

train.head(5)

abstract                 articleID  articleWordCount  \
0      NaN  58def1347c459f24986d7c80               716   
1      NaN  58def3237c459f24986d7c84               823   
2      NaN  58def9f57c459f24986d7c90               575   
3      NaN  58defd317c459f24986d7c95              1374   
4      NaN  58df09b77c459f24986d7ca7               708   

                                byline documentType  \
0  By STEPHEN HILTNER and SUSAN LEHMAN      article   
1                      By GAIL COLLINS      article   
2               By THE EDITORIAL BOARD      article   
3                    By MICHAEL POWELL      article   
4                         By DEB AMLEN      article   

                                            headline  \
0  Finding an Expansive View  of a Forgotten Peop...   
1                  And Now,  the Dreaded Trump Curse   
2              Venezuela’s Descent Into Dictatorship   
3              Stain Permeates Basketball Blue Blood   
4                          Taking Things for Granted   

                                            keywords  multimedia    newDesk  \
0  ['Photography', 'New York Times', 'Niger', 'Fe...           3    Insider   
1  ['United States Politics and Government', 'Tru...           3       OpEd   
2  ['Venezuela', 'Politics and Government', 'Madu...           3  Editorial   
3  ['Basketball (College)', 'University of North ...           3     Sports   
4                              ['Crossword Puzzles']           3      Games   

   printPage              pubDate         sectionName  \
0          2  2017-04-01 00:15:41             Unknown   
1         23  2017-04-01 00:23:58             Unknown   
2         22  2017-04-01 00:53:06             Unknown   
3          1  2017-04-01 01:06:52  College Basketball   
4          0  2017-04-01 02:00:14             Unknown   

                                             snippet              source  \
0  One of the largest photo displays in Times his...  The New York Times   
1                  Meet the gang from under the bus.  The New York Times   
2  A court ruling annulling the legislature’s aut...  The New York Times   
3  For two decades, until 2013, North Carolina en...  The New York Times   
4  In which Howard Barkin and Will Shortz teach u...  The New York Times   

  typeOfMaterial                                             webURL  
0           News  https://www.nytimes.com/2017/03/31/insider/nig...  
1          Op-Ed  https://www.nytimes.com/2017/03/31/opinion/and...  
2      Editorial  https://www.nytimes.com/2017/03/31/opinion/ven...  
3           News  https://www.nytimes.com/2017/03/31/sports/ncaa...  
4           News  https://www.nytimes.com/2017/03/31/crosswords/...

<font color='blue'>Initial goals: 

-Make sure the contents of each field are the correct type and have no missing data (i.e. scrub the 'NaN' from the 'abstract' field)

-Make sure that the data comes properly tokenized

-Convert all words to lowercase (to avoid confusion between uppercase and lowercase versions of the same word)

Several of these data columns (articleID, articleWordCount, multimedia, printPage) contain only integers or single lowercase words.</font> 

In [41]:
#nltk.download('wordnet')
for column in train:
    print(train[column].get_dtype_counts())

object    1
dtype: int64
object    1
dtype: int64
int64    1
dtype: int64
object    1
dtype: int64
object    1
dtype: int64
object    1
dtype: int64
object    1
dtype: int64
int64    1
dtype: int64
object    1
dtype: int64
int64    1
dtype: int64
object    1
dtype: int64
object    1
dtype: int64
object    1
dtype: int64
object    1
dtype: int64
object    1
dtype: int64
object    1
dtype: int64


<font color='blue'>From the above code, the only integer columns are 2, 7 and 9. The rest are string columns and need to be converted to lowercase. </font> 

In [54]:
sampleSize = 4000000
train = pd.read_csv(r"C:\Users\msteele9\Documents\Springboard\Springboard\Data\ArticlesApril2017.csv", header=0, nrows=sampleSize)
#train.head(5)

wn = nltk.WordNetLemmatizer()

nonstrings = [2, 7, 9]


train= train.astype(str)
train.fillna(0)

def clean_articles(doc):
    for index, column in enumerate(doc):
        if index in nonstrings:
            doc[column] = doc[column].astype(str)
            continue
        doc[column] = doc[column].str.replace('[^\w\s]','')
        doc[column] = doc[column].str.lower()
        #doc[column] = doc[column].str.strip()
        doc[column] = doc[column].replace(np.nan, ' ', regex=True)
        doc[column].apply(nltk.word_tokenize)
        doc[column].apply(lemmatize_text)
        doc[column] = [token for token in doc[column] if token not in stop_words]
    return doc

def lemmatize_text(tokenized_text):
    text = [wn.lemmatize(word) for word in tokenized_text]
    return text

clean_art = clean_articles(train)

clean_art.head(5)

#print('\n'.join(clean))


abstract                 articleID articleWordCount  \
0      nan  58def1347c459f24986d7c80              716   
1      nan  58def3237c459f24986d7c84              823   
2      nan  58def9f57c459f24986d7c90              575   
3      nan  58defd317c459f24986d7c95             1374   
4      nan  58df09b77c459f24986d7ca7              708   

                                byline documentType  \
0  by stephen hiltner and susan lehman      article   
1                      by gail collins      article   
2               by the editorial board      article   
3                    by michael powell      article   
4                         by deb amlen      article   

                                            headline  \
0  finding an expansive view  of a forgotten peop...   
1                   and now  the dreaded trump curse   
2               venezuelas descent into dictatorship   
3              stain permeates basketball blue blood   
4                          taking things for granted   

                                            keywords multimedia    newDesk  \
0  photography new york times niger ferguson adam...          3    insider   
1  united states politics and government trump do...          3       oped   
2   venezuela politics and government maduro nicolas          3  editorial   
3  basketball college university of north carolin...          3     sports   
4                                  crossword puzzles          3      games   

  printPage          pubDate         sectionName  \
0         2  20170401 001541             unknown   
1        23  20170401 002358             unknown   
2        22  20170401 005306             unknown   
3         1  20170401 010652  college basketball   
4         0  20170401 020014             unknown   

                                             snippet              source  \
0  one of the largest photo displays in times his...  the new york times   
1                   meet the gang from under the bus  the new york times   
2  a court ruling annulling the legislatures auth...  the new york times   
3  for two decades until 2013 north carolina enga...  the new york times   
4  in which howard barkin and will shortz teach u...  the new york times   

  typeOfMaterial                                             webURL  
0           news  httpswwwnytimescom20170331insidernigermigrants...  
1           oped  httpswwwnytimescom20170331opinionandnowthedrea...  
2      editorial  httpswwwnytimescom20170331opinionvenezuelasdes...  
3           news  httpswwwnytimescom20170331sportsncaabasketball...  
4           news  httpswwwnytimescom20170331crosswordstakingthin...

<font color='blue'>To be safe, I want to check for cells that have missing elements. </font> 

In [35]:
print("Any null values left: "),
clean_art.isnull().values.any()

Any null values left: 


False

<font color='blue'>Now we want a separate section to clean the comment files. </font> 

In [55]:
#For all data - train = pd.read_csv("/root/Springboard/Data/cleaning/allComments.csv")

sampleSize = 100000
train = pd.read_csv(r"C:\Users\msteele9\Documents\Springboard\CommentsApril2017.csv", nrows=sampleSize)

train= train.astype(str)
train.fillna(0)
strings = [1, 5, 10, 18, 22, 24, 25, 26, 29, 30, 33]

def lemmatize_text(tokenized_text):
    text = [wn.lemmatize(word) for word in tokenized_text]
    return text

def clean_articles(doc):
    for index, column in enumerate(doc):
        if index in strings:         
            doc[column] = doc[column].str.replace('[^\w\s]','')
            doc[column] = doc[column].str.lower()
            #doc[column] = doc[column].str.strip()
            doc[column] = doc[column].replace(np.nan, '', regex=True)
            doc[column].apply(nltk.word_tokenize)
            doc[column].apply(lemmatize_text)
            doc[column].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop_words)]))
        else:
            doc[column] = doc[column].astype(str)
            continue
    return doc

normalize_corpus = np.vectorize(norm_articles)

clean_comments = clean_articles(train)
#The second command takes awhile to run
clean_comments.head(5)

C:\Users\msteele9\AppData\Local\Programs\Python\Python36\lib\site-packages\IPython\core\interactiveshell.py:3044: DtypeWarning: Columns (25,26) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


approveDate                                        commentBody   commentID  \
0  1491245186  this project makes me happy to be a 30 year ti...  22022598.0   
1  1491188619  stunning photos and reportage infuriating that...  22017350.0   
2  1491188617  brilliant work from conception to execution iv...  22017334.0   
3  1491167820  nyt reporters should provide a contributors li...  22015913.0   
4  1491167815       could only have been done in print stunning   22015466.0   

  commentSequence commentTitle commentType    createDate depth  \
0        22022598        <br/>     comment  1491237056.0     1   
1        22017350          nan     comment  1491180489.0     1   
2        22017334        <br/>     comment  1491179470.0     1   
3        22015913        <br/>     comment  1491150196.0     1   
4        22015466        <br/>     comment  1491147284.0     1   

  editorsSelection parentID parentUserDisplayName    permID  \
0            False      0.0                   nan  22022598   
1            False      0.0                   nan  22017350   
2            False      0.0                   nan  22017334   
3            False      0.0                   nan  22015913   
4            False      0.0                   nan  22015466   

                                              picURL recommendations  \
0  https://graphics8.nytimes.com/images/apps/time...             2.0   
1  https://graphics8.nytimes.com/images/apps/time...             1.0   
2  https://graphics8.nytimes.com/images/apps/time...             3.0   
3  https://graphics8.nytimes.com/images/apps/time...             7.0   
4  http://profile.ak.fbcdn.net/hprofile-ak-snc4/h...             5.0   

  recommendedFlag replyCount reportAbuseFlag sharing    status timespeople  \
0             nan        0.0             nan       0  approved         1.0   
1             nan        0.0             nan       0  approved         1.0   
2             nan        0.0             nan       0  approved         1.0   
3             nan        2.0             nan       0  approved         1.0   
4             nan        0.0             nan       0  approved         1.0   

  trusted  updateDate userDisplayName    userID    userLocation userTitle  \
0     0.0  1491245186       rob gayle  46006296    riverside ca       nan   
1     0.0  1491188619         susan a  29202761              br       nan   
2     0.0  1491188617            meta  63944806      raleigh nc       nan   
3     0.0  1491167820      tom wyrick   1266184    missouri usa       nan   
4     0.0  1491167815     joe sharkey  61121360  tucson arizona       nan   

  userURL inReplyTo                 articleID sectionName  newDesk  \
0     nan         0  58def1347c459f24986d7c80     unknown  insider   
1     nan         0  58def1347c459f24986d7c80     unknown  insider   
2     nan         0  58def1347c459f24986d7c80     unknown  insider   
3     nan         0  58def1347c459f24986d7c80     unknown  insider   
4     nan         0  58def1347c459f24986d7c80     unknown  insider   

  articleWordCount printPage typeOfMaterial  
0            716.0         2           news  
1            716.0         2           news  
2            716.0         2           news  
3            716.0         2           news  
4            716.0         2           news

In [56]:
print("Any null values left: "), print(clean_comments.isnull().values.any())

Any null values left: 
False


(None, None)

In [57]:
# Make sure that I am saving the files where I want
import os
os.chdir(r'C:\Users\msteele9\Documents\Springboard\Springboard\Data')
art_file_name = "cleaned_article_data.csv"
clean_art_csv = clean_art.to_csv(art_file_name, encoding='utf-8', index=False)

In [58]:
com_file_name = "cleaned_comment_data.csv"
clean_com_csv = clean_comments.to_csv(com_file_name, encoding='utf-8', index=False)

<font color='blue'>Here I have saved the cleaned data files to storage. This is a good break point where I can use these files to begin looking at possible ML models. 
    
   In the next section, I will explore the data, looking at features and ways to visualize the feature set. I begin by re-importing the libraries that I will use. That way, this section of the notebook can be reran separately later. I will load the cleaned files in so that I know I am not making any changes to the stored version.</font> 

In [59]:
import pandas as pd
import numpy as np
import nltk
#ntlk.download()
import re
sent_token = nltk.sent_tokenize
import csv  
from nltk import sent_tokenize, word_tokenize, pos_tag
import re
from sklearn.feature_extraction.text import CountVectorizer
wpt = nltk.WordPunctTokenizer()
stop_words = nltk.corpus.stopwords.words('english')

file_path_art = r'C:\Users\msteele9\Documents\Springboard\Springboard\Data\cleaned_article_data.csv'
clean_art = pd.read_csv(file_path_art, index_col = False)
clean_art

file_path_comments = r'C:\Users\msteele9\Documents\Springboard\Springboard\Data\cleaned_comment_data.csv'
clean_comments = pd.read_csv(file_path_comments, index_col = False)

<font color='blue'> In the above cell, I loaded in the cleaned article data. Here I load in the cleaned data and do preprocessing to run the data through a random forest model. 
    
   I want to make sure that my data can actually be loaded and that my simple test will produce results of some sort - this way I know that my data will not break when I try to load it.
   
   My target is the number of recommendations that the comment receives. I want to know how well a simple model can predict the number of recommendations a comment will receive when given the rest of that comment's data.
</font> 

In [60]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()

features = clean_comments.columns.tolist()
output = 'recommendations'
features.remove('recommendations')

for column in clean_comments.columns:
    clean_comments[column] = clean_comments[column].astype(str)
    if clean_comments[column].dtype == type(object):
        clean_comments[column] = le.fit_transform(clean_comments[column])

#print(features)

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

clean_comments.astype(int)
clean_comments.head(5)

#clean_comments.dtypes

approveDate  commentBody  commentID  commentSequence  commentTitle  \
0         5675        82175      10692            10692             0   
1         3457        67130       7166             7166             1   
2         3456        11257       7160             7160             0   
3         3292        54786       6606             6606             0   
4         3291        14150       6385             6385             0   

   commentType  createDate  depth  editorsSelection  parentID  \
0            0       10239      0                 0         0   
1            0        6940      0                 0         0   
2            0        6934      0                 0         0   
3            0        6388      0                 0         0   
4            0        6176      0                 0         0   

   parentUserDisplayName  permID  picURL  recommendations  recommendedFlag  \
0                   5571   11368      98              247                0   
1                   5571    7306      98                1                0   
2                   5571    7301      98              387                0   
3                   5571    6844      98              745                0   
4                   5571    6688      73              592                0   

   replyCount  reportAbuseFlag  sharing  status  timespeople  trusted  \
0           0                0        0       0            1        0   
1           0                0        0       0            1        0   
2           0                0        0       0            1        0   
3          15                0        0       0            1        0   
4           0                0        0       0            1        0   

   updateDate  userDisplayName  userID  userLocation  userTitle  userURL  \
0        5887            21206    9821          5970          0        1   
1        3648            23831    5481           935          0        1   
2        3647            16895   19981          5828          0        1   
3        3473            24897     659          4372          0        1   
4        3472            12207   17732          7277          0        1   

   inReplyTo  articleID  sectionName  newDesk  articleWordCount  printPage  \
0          0          0           26        9               222         12   
1          0          0           26        9               222         12   
2          0          0           26        9               222         12   
3          0          0           26        9               222         12   
4          0          0           26        9               222         12   

   typeOfMaterial  
0               5  
1               5  
2               5  
3               5  
4               5

In [61]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import GridSearchCV
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, log_loss
from sklearn.model_selection import GridSearchCV

rf = RandomForestClassifier()

param = {'n_estimators': [5,10,100, 200, 500, 1000],
         'max_depth': [5, 10, 30, 100, 1000] }

gs=GridSearchCV(rf, param, cv=5, n_jobs=-1)


In [63]:
pd.DataFrame(gs_fit.cv_results_).sort_values(by=['rank_test_score']).head(5)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
4       0.687200      0.075735         0.047400        0.002417   
3       0.266600      0.035364         0.018600        0.001200   
5       1.401199      0.168370         0.095801        0.004996   
0       0.008200      0.000748         0.001802        0.000400   
2       0.131799      0.021302         0.010000        0.000894   

  param_max_depth param_n_estimators                                  params  \
4               5                500   {'max_depth': 5, 'n_estimators': 500}   
3               5                200   {'max_depth': 5, 'n_estimators': 200}   
5               5               1000  {'max_depth': 5, 'n_estimators': 1000}   
0               5                  5     {'max_depth': 5, 'n_estimators': 5}   
2               5                100   {'max_depth': 5, 'n_estimators': 100}   

   split0_test_score  split1_test_score  split2_test_score  split3_test_score  \
4           0.061594           0.147059           0.135870           0.180233   
3           0.065217           0.142157           0.125000           0.174419   
5           0.072464           0.142157           0.125000           0.180233   
0           0.086957           0.132353           0.130435           0.139535   
2           0.050725           0.137255           0.135870           0.174419   

   split4_test_score  mean_test_score  std_test_score  rank_test_score  
4           0.134146            0.125        0.042060                1  
3           0.146341            0.124        0.039246                2  
5           0.115854            0.122        0.036759                3  
0           0.115854            0.118        0.020420                4  
2           0.128049            0.118        0.044132                4

<font color='blue'> A quick GridSearch with a random forest classifier was able run on my data set and produce an accuracy score. The highest-performing model (mean score 0.125) used a high number of trees (500) and a low max tree depth (5). The models with more trees performed better, but the results are clumped fairly close together. All of the top performers used the minimum allowed tree depth.
    
   My goal was to confirm that I can run ML algorithms on my data set and get sensible results; this test seems to verify this.
</font> 